[Profiling] Allow wildcards for custom indices

danielmitterdorfer opened the pull request:
Several profiling APIs allow to specify custom event indices with the `indices` parameter. So far this parameter has allowed only one index and also disallowed wildcards. With this commit we extend this parameter to support multiple indices and each index can use wildcards.

danielmitterdorfer committed:
[Profiling] Allow wildcards for custom indices

Several profiling APIs allow to specify custom event indices with the
`indices` parameter. So far this parameter has allowed only one index
and also disallowed wildcards. With this commit we extend this parameter
to support multiple indices and each index can use wildcards.

elasticsearchmachine commented:
Pinging @elastic/obs-knowledge-team (Team:obs-knowledge)

danielmitterdorfer commented:
Note: I'm labeling this as non-issue as the respective API is exclusive to Kibana and it is not yet in use.

danielmitterdorfer commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/GetStackTracesRequest.java:
```diff
@@ -32,7 +35,7 @@
  * A request to get profiling details
  */
-public class GetStackTracesRequest extends ActionRequest implements IndicesRequest {
+public class GetStackTracesRequest extends ActionRequest implements IndicesRequest.Replaceable {
```
Implementing `IndicesRequest.Replaceable` indicates to infrastructure in Elasticsearch that we want to support wildcards, see also https://github.com/elastic/elasticsearch/blob/510c8515abecf5ca23154e25a72247329e8aa311/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L136-L146

rockdaboot commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/GetStackTracesRequest.java:
```diff
@@ -221,6 +219,28 @@ public void parseXContent(XContentParser parser) throws IOException {
+                );
+            }
+        }
+        return indices.toArray(new String[0]);
```
Should the String array be pre-allocated with `indices.size()`?

danielmitterdorfer replied:
That's actually the preferred style in Java, see also https://github.com/JetBrains/intellij-community/blob/master/java/java-impl/src/inspectionDescriptions/ToArrayCallWithZeroLengthArrayArgument.html for the rationale.

rockdaboot commented on x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/GetStackTracesRequest.java:
```diff
@@ -330,7 +350,7 @@ public String[] indices() {
-            indices.add(this.indices);
+            indices.addAll(List.of(this.indices));
         }
         return indices.toArray(new String[0]);
```
Maybe use pre-allocation here as well?

danielmitterdorfer replied:
That's actually the preferred style in Java, see also https://github.com/JetBrains/intellij-community/blob/master/java/java-impl/src/inspectionDescriptions/ToArrayCallWithZeroLengthArrayArgument.html for the rationale.

rockdaboot reviewed (approved):
LGTM (left some minor comments)

danielmitterdorfer committed:
Allow to distinguish whether indices are user-provided

As we set indices now internally in all cases due to wildcard
resolution, we need a means to distinguish whether `indices` have been
provided originally via the API or whether Elasticsearch has set them
internally after wildcard resolution.

danielmitterdorfer committed:
Checkstyle

danielmitterdorfer committed:
Empty array
